nixos/transmission: handle watch-dir

doronbehar opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Noticed I got errors with watch-dir ever since hardening. cc @ju1m / @Lassulus .

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

doronbehar committed:
nixos/transmission: handle watch-dir as incomplete-dir

`watch-dir` was neglected after #92106 - this change makes using this
setting work.

ju1m commented:
@doronbehar Indeed support for `watch-dir` is missing, thanks. I cannot test this PR right now, but I suppose that:
- Because the user's settings may not define it, a default value must be added for `watch-dir-enabled`.
- Also, a sensible default like `"${cfg.home}/Torrents"` or else, may be set for `watch-dir`.
- Maybe using `BindReadOnlyPaths=` (instead of `BindPaths=`) would be enough for `watch-dir` to work correctly.
- A rule `r ${cfg.settings.watch-dir},` guarded by `cfg.settings.watch-dir-enabled` must also be added in `transmission-daemon`'s AppArmor profile.

doronbehar commented:
> Because the user's settings may not define it, a default value must be added for watch-dir-enabled
> Also, a sensible default like `"${cfg.home}/Torrents"` or else, may be set for `watch-dir`.

There's an upstream default for `watch-dir` which is `"${cfg.home}/watchdir"` and it's `false` by default. I copied most of what you did with `incomplete-dir` for `watch-dir`, according to upstream's defaults, so I hope it should be OK now.

> Maybe using BindReadOnlyPaths= (instead of BindPaths=) would be enough for watch-dir to work correctly.

No, because after Transmission is adding a torrent, it renames the file with the `.added` suffix, so it should need write permissions as well.

> A rule r ${cfg.settings.watch-dir}, guarded by cfg.settings.watch-dir-enabled must also be added in transmission-daemon's AppArmor profile.

Agreed. I added this plus other checks based on `incomplete-dir`.

ju1m commented:
@doronbehar Great :) I still can't test it, but it looks like there is a missing `''}` in the AppArmor profile there: https://github.com/NixOS/nixpkgs/commit/42dd740ffe0230a91ad22f39932d497b6d9f1764#diff-3b9521332ddaab17070757e70eb2ff9bR421

doronbehar commented:
Correct. I fixed that and tested the `transmission.nix` test and it succeeded.

ju1m commented:
Well then, LGTM, thanks for this PR @doronbehar :)
